### PR TITLE
[core] Support magic links in `SignInPage`

### PR DESCRIPTION
**Pull request by bharatkashyap**
- Closes https://github.com/mui/mui-toolpad/issues/3990

Docs preview: https://deploy-preview-4085--mui-toolpad-docs.netlify.app/toolpad/core/react-sign-in-page/#magic-link

### CODE REVIEW

File: docs/data/toolpad/core/components/sign-in-page/sign-in-page.md
Comment (apedroferreira) [inline docs/data/toolpad/core/components/sign-in-page/sign-in-page.md:51]: Should there be a default success alert that would already show in this demo?

File: docs/data/toolpad/core/components/sign-in-page/sign-in-page.md
Comment (apedroferreira) [inline docs/data/toolpad/core/components/sign-in-page/sign-in-page.md:66]: I guess they don't really "need" to but it's a way that users can do it, right? So maybe we can say they "can" add that code instead, as this is an example of how they can do it with Auth.js?

File: docs/data/toolpad/core/components/sign-in-page/sign-in-page.md
Comment (apedroferreira) [inline docs/data/toolpad/core/components/sign-in-page/sign-in-page.md:93]: Would a more specific name like `successAlert` or `successMessage` be better?

File: examples/core-auth-nextjs-email/src/app/auth/signin/actions.ts
Comment (apedroferreira) [inline examples/core-auth-nextjs-email/src/app/auth/signin/actions.ts:23]: I guess they don't provide types that we can use to type this error with the `digest`?

File: docs/data/toolpad/core/components/sign-in-page/sign-in-page.md
Comment (apedroferreira) [inline docs/data/toolpad/core/components/sign-in-page/sign-in-page.md:89]: Not super clean that this is how "success" works (by catching an error) but if that's just how Auth.js does it I guess there's nothing we can do?
Ideally this code would be easier to follow.

**Review by apedroferreira — APPROVED** [review]
Cool, seems to work well, the demos look clean!
I wrote down some thoughts and ideas.

File: docs/data/toolpad/core/components/sign-in-page/sign-in-page.md
Comment (bharatkashyap) [reply docs/data/toolpad/core/components/sign-in-page/sign-in-page.md:92]: This is how Next.js + Auth.js server compoents redirects work - throwing a `NEXT_REDIRECT` which we must catch, I'm basing this on https://github.com/vercel/next.js/issues/49298#issuecomment-1542055642

File: examples/core-auth-nextjs-email/src/app/auth/signin/actions.ts
Comment (bharatkashyap) [reply examples/core-auth-nextjs-email/src/app/auth/signin/actions.ts:23]: I agree that this is quite fragile, but I couldn't find any way to detect exclusively that the redirect is to `verify-request` using an Error type. I'll investigate a bit more

File: docs/data/toolpad/core/components/sign-in-page/sign-in-page.md
Comment (apedroferreira) [reply docs/data/toolpad/core/components/sign-in-page/sign-in-page.md:96]: Got it, just a thought, I think `error` shouldn't be renamed so we can keep `success` if it's more consistent with that.

File: examples/core-auth-nextjs-email/src/app/auth/signin/actions.ts
Comment (apedroferreira) [reply examples/core-auth-nextjs-email/src/app/auth/signin/actions.ts:23]: Ok, no big deal if you can't find it, was just wondering.

File: examples/core-auth-nextjs-email/src/auth.ts
Comment (Janpot) [inline examples/core-auth-nextjs-email/src/auth.ts:71]: This shouldn't throw in production and log in development. How do you feel about making this drastically simpler. Just

```tsx
if (!process.env.GITHUB_CLIENT_ID) {
  console.warn('Missing environment variable "GITHUB_CLIENT_ID"')
}
```

Just keeping it plain and simple, very readable, no extra logic, no opinions, no colors in the messages. We could let end-users make their own choices about how fancy they want this to be.

File: examples/core-auth-nextjs-email/src/auth.ts
Comment (Janpot) [inline examples/core-auth-nextjs-email/src/auth.ts:50]: This doesn't seem necessary here https://github.com/mui/toolpad/blob/4e00a5fc342babd9c900c24cc41359d5a3a985ba/playground/nextjs/src/auth.ts#L29-L35

```suggestion
      id: providerData.id,
```

File: docs/data/toolpad/core/components/sign-in-page/sign-in-page.md
Comment (Janpot) [inline docs/data/toolpad/core/components/sign-in-page/sign-in-page.md:54]: This page still gives me the impression that Toolpad core is about providing a full integration end-to-end, while in reality it's just about providing UI. I feel like it would make more sense to come from the angle that this component aims to provide UI for the auth.js Nodemailer provider.

File: examples/core-auth-nextjs-email/src/auth.ts
Comment (bharatkashyap) [reply examples/core-auth-nextjs-email/src/auth.ts:71]: I deployed one of the examples on a personal server to test the integration a few days ago and found the throwing in production very helpful; also personally the colours help make it more readable. Since this is personal opinion, I'm open to removing it if you think that makes more sense

File: docs/data/toolpad/core/components/sign-in-page/sign-in-page.md
Comment (bharatkashyap) [reply docs/data/toolpad/core/components/sign-in-page/sign-in-page.md:54]: I can reword the warning and the introduction text of this section like the following:


:::warning
To use magic links, apart from configuring an authentication provider like Auth.js, you must also configure a database and pass in the required environment variables to send emails. See more details in the Auth.js docs on [database setup for email](https://authjs.dev/getting-started/authentication/email) and [Nodemailer configuration](https://authjs.dev/getting-started/providers/nodemailer/).
:::

You can use the `SignInPage` component to quickly setup a sign-in page for the `nodemailer` provider offered by Auth.js which sends one-time verification links to the user's email address.

Comment (Janpot) [reply docs/data/toolpad/core/components/sign-in-page/sign-in-page.md:54]: I'd expect something along the lines of

The Signin page component supports magic links. In order to enable this you'll have to set up a provider such as Auth.js NodeMailer. Please note that such provider requires persistence.

File: examples/core-auth-nextjs-email/src/auth.ts
Comment (Janpot) [reply examples/core-auth-nextjs-email/src/auth.ts:71]: We're not (yet) in the business of configuration management. Until we are, I feel like it makes most sense to be very bare bones so that users see in the blink of an eye what this code does and how they can replace it with the configuration management of their choice.

File: examples/core-auth-nextjs-email/src/auth.ts
Comment (Janpot) [reply examples/core-auth-nextjs-email/src/auth.ts:50]: I believe that shouldn't be necessary either. At most I'd expect the export to be typed

```tsx
export const providerMap: AuthProvider[] = // ...
```

File: docs/data/toolpad/core/components/sign-in-page/sign-in-page.md
Comment (Janpot) [inline docs/data/toolpad/core/components/sign-in-page/sign-in-page.md:53]:
```suggestion
The `SignIn` page component supports magic links. To enable this, you have to set up a provider such as Auth.js NodeMailer. See more details in the Auth.js docs on [database setup for email](https://authjs.dev/getting-started/authentication/email) and [Nodemailer configuration](https://authjs.dev/getting-started/providers/nodemailer/).
```

File: docs/data/toolpad/core/components/sign-in-page/sign-in-page.md
Comment (Janpot) [inline docs/data/toolpad/core/components/sign-in-page/sign-in-page.md:241]: Address vale comment

```suggestion
Besides the `SignInPage` , the team is planning work on several other components that enable new workflows such as sign up and password reset.
```

Perhaps we need links to issues?

File: docs/data/toolpad/core/components/sign-in-page/sign-in-page.md
Comment (bharatkashyap) [reply docs/data/toolpad/core/components/sign-in-page/sign-in-page.md:51]: We have a separate demo on the Alerts just below this one, so I think this one is okay? If you think strongly that we should merge them, I can do that

**Comment by jakobmerrild** [comment]
@bharatkashyap I was looking for any open issues w.r.t. supporting one-time passwords when using the `credentials` provider and found [your issue](https://github.com/mui/toolpad/issues/3990). However, looking at the code in this PR it doesn't seem to solve that part of the issue.
Do you have any plans to include support for one-time passwords in the credentials provider with this PR?

**Comment by bharatkashyap** [comment]
> @bharatkashyap I was looking for any open issues w.r.t. supporting one-time passwords when using the `credentials` provider and found [your issue](https://github.com/mui/toolpad/issues/3990). However, looking at the code in this PR it doesn't seem to solve that part of the issue. Do you have any plans to include support for one-time passwords in the credentials provider with this PR?

@jakobmerrild Not with this PR (I should edit the title of that issue) but the plan is to implement support for OTPs as an immediate follow up of this PR, including adding an example app. I'll create a separate issue: https://github.com/mui/toolpad/issues/4292

**Comment by jakobmerrild** [comment]
Sounds great!

**Comment by jakobmerrild** [comment]
Actually, looking at the new issue it sounds like you are planning to support one time codes sent to the user's email.

What I'm looking for is 2FA support via one-time passwords, basically an additional optional input field on the `SignInPage`.

**Comment by jakobmerrild** [comment]
Thanks for letting me know. I have opened [an issue](https://github.com/mui/toolpad/issues/4293). In the mean time is there a way for me to create my *own* sign-in page and have it integrate with the `Account` component?

**Comment by bharatkashyap** [comment]
> Thanks for letting me know. I have opened [an issue](https://github.com/mui/toolpad/issues/4293). In the mean time is there a way for me to create my _own_ sign-in page and have it integrate with the `Account` component?

Thanks for the detailed issue, I'll add it to the consideration for our future versions. 

For the meantime, you could use the `Account` component separately like the demos in this page: https://mui.com/toolpad/core/react-account/ while building your own component for the sign in page.

**Comment by jakobmerrild** [comment]
Thanks! I managed to get it to work using your example and the basic page from `authjs`. Should then be able to just create a new custom page and pass it into the `pages` prop for the `NextAuth` I believe.
Though I am curious why your [example](https://github.com/mui/toolpad/blob/949b6bf22a810269fc1cd8b431845f65165e1dfd/examples/core-auth-nextjs/src/app/layout.tsx#L7) uses the `signIn` and `signOut` functions from `next-auth/react` instead the ones exported from `auth.ts`.

**Comment by bharatkashyap** [comment]
> Though I am curious why your [example](https://github.com/mui/toolpad/blob/949b6bf22a810269fc1cd8b431845f65165e1dfd/examples/core-auth-nextjs/src/app/layout.tsx#L7) uses the `signIn` and `signOut` functions from `next-auth/react` instead the ones exported from `auth.ts`.

This is because `SignInPage` is a client component - see https://authjs.dev/getting-started/migrating-to-v5#details

**Comment by jakobmerrild** [comment]
Gotcha! Thanks for all the help and sorry for spamming your PR 🙈

File: packages/toolpad-core/src/SignInPage/SignInPage.test.tsx
Comment (apedroferreira) [inline packages/toolpad-core/src/SignInPage/SignInPage.test.tsx:60]: I guess that none of those usual methods like `toHaveBeenLastCalledWith` works for these checks? Or would it?

File: packages/toolpad-core/src/SignInPage/SignInPage.tsx
Comment (apedroferreira) [inline packages/toolpad-core/src/SignInPage/SignInPage.tsx:561]: Maybe try to use theme units if possible unless we really need the 12px specifically here?

File: packages/toolpad-core/src/SignInPage/SignInPage.tsx
Comment (apedroferreira) [inline packages/toolpad-core/src/SignInPage/SignInPage.tsx:268]: A `.filter` here would probably be cleaner, before using `.map`?

File: packages/toolpad-core/src/SignInPage/SignInPage.tsx
Comment (apedroferreira) [inline packages/toolpad-core/src/SignInPage/SignInPage.tsx:269]: You can also just create a single boolean variable for this provider id check, might end up being useful for other situations later too.

**Review by apedroferreira — APPROVED** [review]
No major changes since last time I approved, right? Let me know if there's anything I should check more in specific.
LGTM, added some comment that sound worth doing before merging, but nothing blocking!

**Pull request merged**
